lucene/core/src/java/org/apache/lucene/search/TopDocsCollector.java

```diff
@@ -136,14 +136,21 @@ public TopDocs topDocs(int start, int howMany) {
     // pq.size() or totalHits.
     int size = topDocsSize();
 
-    // Don't bother to throw an exception, just return an empty TopDocs in case
-    // the parameters are invalid or out of range.
-    // TODO: shouldn't we throw IAE if apps give bad params here so they dont
-    // have sneaky silent bugs?
-    if (start < 0 || start >= size || howMany <= 0) {
+
+    if (start < 0 || start > size) {
```

Comment:
I don't think we should validate `start` against `size` since `size` depends on the number of matches, and callers have no clue about its value at this point? Otherwise we are making this method almost impossible to use?

The thing that we should validate is that `start + howMany` is less than or equal to the `numHits` parameter (which doesn't seem to be passed to TopDocsCollector by its sub classes at the moment, but we could)?

Reply:
@jpountz That can allow potential abuse by passing in Integer.MAX_VALUE for numHits, and a value of start > size. With this check gone, we would not be able to detect the case.

Should we keep the existing one and add the one you proposed?